pkg/scrape/scrape.go

```diff
@@ -0,0 +1,193 @@
+	res, err := http.Get(url)
+	if err != nil {
+		return nil, err
+	}
+	if res.StatusCode != 200 {
+		return nil, err
+	}
+	return res, err
+}
+
+func GetResultsForLetter(letter string) (types.SearchResults, error) {
+	url := fmt.Sprintf("%s%s", LetterBase, letter)
+	res, err := DownloadPage(url)
+	defer func(Body io.ReadCloser) {
+		err := Body.Close()
+		if err != nil {
+			panic(err)
+		}
+	}(res.Body)
+	if err != nil {
+		return nil, err
+	}
+	doc, err := goquery.NewDocumentFromReader(res.Body)
+	if err != nil {
+		return nil, err
+	}
+	results := make(types.SearchResults)
+	doc.Find("#EchoTopic p[align='left'] a").Each(func(i int, s *goquery.Selection) {
+		title := s.Text()
+		results[title] = "#"
+		trackUrl, exists := s.Attr("href")
+		if exists {
+			results[title] = trackUrl
+		} else {
+			results[title] = "#"
+		}
+	})
+	return results, nil
+}
+
+func RetrieveAlbum(path string) (types.Album, error) {
+	var album types.Album
+	slugBits := strings.Split(path, "/")
+	album.Slug = slugBits[len(slugBits)-1]
+	albumUrl := fmt.Sprintf("%s%s", util.SiteBase, path)
+	res, err := DownloadPage(albumUrl)
+	defer func(Body io.ReadCloser) {
+		err := Body.Close()
+		if err != nil {
+			panic(err)
+		}
+	}(res.Body)
+	if err != nil {
+		return album, err
+	}
+	doc, err := goquery.NewDocumentFromReader(res.Body)
+	if err != nil {
+		return album, err
+	}
+	metadata := doc.Find("#EchoTopic p[align='left'] b")
+	if metadata.Length() == 5 {
+		album.FlacAvailable = true
+	}
+	metadata.Each(func(i int, s *goquery.Selection) {
+		if i == 0 {
+			album.Name = s.Text()
+		}
+		if i == 1 {
+			album.FileCount, err = strconv.Atoi(s.Text())
+			if err != nil {
+				album.FileCount = 0
+			}
+		}
+		if i == 2 {
+			album.MP3FileSize = s.Text()
+		}
+		if i == 3 && album.FlacAvailable {
+			album.FlacFileSize = s.Text()
+		}
+	})
+	flacLabel := ""
+	if album.FlacAvailable {
+		flacLabel = "[FLAC]"
+	}
+	pterm.Success.Printfln(
+		"Found %s (%d tracks) %s %s",
+		album.Name,
+		album.FileCount,
+		"[MP3]",
+		flacLabel)
+	pterm.Warning.Println("Searching for track locations up front. This may take seconds or minutes depending on album length!")
+	_ = doc.Find("#EchoTopic table:not(#songList) tr div a").Each(func(i int, s *goquery.Selection) {
+		coverUrl, exists := s.Attr("href")
+		if exists {
+			// TODO: Use proper escaping. Tried stdlib but it escaped everything
+			coverUrl = strings.ReplaceAll(coverUrl, " ", "%20")
+			album.Covers = append(album.Covers, coverUrl)
+		}
+	})
+	songMeta := make(map[int]string)
+	_ = doc.Find("#songlist_header th").Each(func(i int, s *goquery.Selection) {
+		header := strings.TrimSpace(s.Text())
+		if header == "CD" {
+			songMeta[i] = "CD"
+		}
+		if header == "#" {
+			songMeta[i] = "TrackNumber"
+		}
+		if header == "Song Name" {
+			songMeta[i] = "SongName"
+		}
+		if header == "MP3" {
+			songMeta[i] = "TrackLength"
+			songMeta[i+1] = "MP3FileSize"
+		}
+		if header == "FLAC" {
```

Comment:
Some albums have `WAV`, `OGG`, and `M4A` that might be worth tracking as well.

Reply:
Yeah, eventually I'll add those but currently it just defaults to `MP3` so even tracking FLAC is a bit pointless. It needs a bit more investigation into how the page layouts look for those other formats.